### ISSUE TITLE
Async: detecting async methods from an interface

### ISSUE THREAD

**Issue by mcquiggd** [Pro-Coded/Polly.Contrib.Decorator#1]
Roslyn doesn't _seem_ to provide the ability to detect if a method is async, from an interface.

Attempting to infer this from a name is problematic (not everyone uses the suffix Async).

Relying on the return type is also not reliable.

More investigation needed

**Comment by mcquiggd** [comment]
For reference:

https://stackoverflow.com/questions/44999586/how-to-identify-if-the-method-implementation-is-marked-as-async-can-be-called

**Comment by reisenberger** [comment]
My answer would have been much the same as the excellent contribs to your StackOverflow q, with the addition that the existence of `async void` methods doesn't affect the conclusion.

**TL;DR** 

+ A method _can_ be `await`ed if-and-only-if its return type is `await`able (this is sufficiently determining).  
+ Whether the `async` keyword appears in a method implementation is _not_ sufficiently determining for whether a call to it can or should use `await`. 
+ `async void` methods do not change the above conclusions; an `async void` method does not run any less asynchronously for the fact that you cannot call it with `await`.

Whether a method call _**should**_ be awaited by `Polly.Contrib.Decorator`, is the subject of issue #2. 
  

#### (1) `async` is not formally part of a method signature; but even if it was, is not sufficient to determine whether a method call could be `await`-ed 

`async` is not formally part of a method signature.  So `async` is not found in interfaces.  So you can't determine from an interface, if the interface's original author intended method implementations to be written with the `async` keyword or called with `await`.

However, whether the called method is written with the `async` keyword is not actually the determining (/even a sufficiently determining) factor, for whether a method could/should be called with `await`.   There are valid cases for writing methods without the `async` keyword which return `await`ables:

[a] [async-await eilision](https://blog.stephencleary.com/2016/12/eliding-async-await.html), as [used extensively by Polly](https://github.com/App-vNext/Polly/blob/master/src/Polly.Shared/PolicyAsync.cs#L27-L252)  
[b] an interface mostly used for implementations over I/O, hence declared with `await`able return types, but for which you might also want sometimes to write an in-memory (so synchronous) implementation/interceptor.  [Discussion of common case: sync in-memory cache around async I/O](http://blog.i3arnon.com/2015/11/30/valuetask/)
[c] for test purposes, stubbing out some asynchronous dependency with an in-memory (sync) stub

(It's actually good that `async` isn't part of the method signature, because it permits us occasionally to fulfil with sync implementations, as above.) 

#### (2) anything that returns a `Task`, `Task<TResult>` or [has a suitable `GetAwaiter()` method](https://blogs.msdn.microsoft.com/pfxteam/2011/01/13/await-anything/), can be awaited

The only thing uniquely determining whether a method call _can_ be `await`-ed is whether its return type is `await`able.  

Whether we _should_ call all such methods with `await` when wrapping with `Polly.Contrib.Decorator`, is the subject of issue #2.

#### (3) `async void` methods do not invalidate the above conclusion

This addresses `async void` methods at length, because of the comment in the SO question that the return type of the method might be insufficient, because `void` could not be distinguished (in the interface) from `async void`.

The starting point is that `async void` methods _cannot_ be awaited, for the simple reason that, although written with the `async` keyword, they do not return any type that can be `await`-ed.  

Does that invalidate our conclusion (2)? (that we can uniquely use whether a method returns an `await`able to determine how to call it).  Are we _losing some capability to run `async void` methods asynchronously_, because we can't `await` them?  

Put concretely: say the method behind the interface is `async void Foo()` but all we know about it from the interface is that it is `void Foo()`.  Are we losing some ability for `Foo()` to run asynchronously if we only invoke it `Foo()`?

The answer is no, because of the way `async` methods operate.  `async void` methods behave just as any `async Task/Task<T>` method called with `await`: they run synchronously until their first internal `await`; they then return (`void`, or a `Task` representing their promise to complete), and schedule the remainder of the called method (the part after the first `await`) as a continuation.  That continuation is the part which will run asynchronously, after the `await`ed thing completes.  (That's rather a condensed description, but this is widely blogged; [example discussion](https://stackoverflow.com/questions/33731376/why-does-an-async-void-method-run-synchronously).) 

In other words: The determining factor that some portion of an `async void` method will run asynchronously is not that it is called with `await`, but that, in _its_ body, it has an `await` with some meaningful work after it. 

#### Another angle on `async void`  

Since the essential q (for the purposes of `Polly.Contrib.Decorator`) is how should we _call_ a wrapped method, we can run an alternative thought experiment around `async void` methods.   What if we could (somehow) determine that a `void` method behind an interface had been declared `async void`?  Would we call it any differently?

Back to our example, `async void Foo()`.  What are our choices?  We could `Foo()` directly or `Task.Run(() => Foo())`.

As Stephen Cleary covers, [a library shouldn't use `Task.Run()` on a caller's behalf](https://blog.stephencleary.com/2013/11/taskrun-etiquette-examples-dont-use.html).  Doing so means the library is taking a choice that work should be offloaded onto a background thread, denying the caller that choice.  (_Note_: per above discussion about how `async void` methods operate, this only applies to the work up to the first `await` within the called method anyway.)   

So: even if we could know the method behind `void Foo()` in the interface was `async void Foo()`, `Polly.Contrib.Decorator` should still only call `Foo()` anyway.  If the user wants to immediately offload the work onto a different thread (eg they want to offload it off the GUI thread), they would (before `Polly.Contrib.Decorator` was introduced) be calling that interface method with `Task.Run(() => ...)` anyway.  We don't need to add an extra one.

This conforms with a principle Polly follows: it should have least effect (apart from the declared intended effects) on how users' delegates are run.  

#### Wrapping up

Key for all the above is that the `async` keyword _doesn't_ (in and of itself) _make a method run asynchronously_ or even concurrently, so is not the clincher.  The `async` keyword just permits the compiler to chop a method up into a series of chunks at `await` statements; chunks 2..n of such a method run (asynchronously; at a different time) as continuations after the preceding `await`ed call completes.  The caller is (except with `async void` methods) returned a `Task` which is a 'promise' that will complete when the `await`ed method completes.  

The fact of being returned that `Task` (or something else implementing `GetAwaiter()`) is the determining factor for whether it can be called with `await`: if the returned type of the method implements this awaitable pattern, it can be awaited.

The existence of async/await elision and the sync-cache-over-async-io pattern in particular, demonstrate that the return type of the called method is key, not whether the implementation used the `async` keyword.

**Comment by mcquiggd** [comment]
As discussed on the Polly Slack channel, at this point in time, I have decided to test the return type of each method, and not test for `GetAwaiter`. The current solution is flexible, and it is perfectly possible for developers to edit the generated code as they see fit.

I have implemented two options, to cover both scenarios where `async `/ `await `may be encountered:

`Implement Interface with Polly`

`Implement Interface with Polly, including async await`

**Comment by reisenberger** [comment]
SGTM.  A simple flexible option and easily comprehensible to users

**Comment by mcquiggd** [comment]
@reisenberger  - I have updated the readme, when time permits I will perhaps include further details, perhaps portions of your posts, and links to Steven Cleary's article on the subject...
